osu.Game/Rulesets/Difficulty/Skills/Skill.cs

```diff
@@ -1,150 +1,47 @@
 
-using System;
 using System.Collections.Generic;
-using System.Linq;
 using osu.Game.Rulesets.Difficulty.Preprocessing;
 using osu.Game.Rulesets.Difficulty.Utils;
 using osu.Game.Rulesets.Mods;
 
 namespace osu.Game.Rulesets.Difficulty.Skills
 {
     /// <summary>
-    /// Used to processes strain values of <see cref="DifficultyHitObject"/>s, keep track of strain levels caused by the processed objects
-    /// and to calculate a final difficulty value representing the difficulty of hitting all the processed objects.
+    /// A bare minimal abstract skill for fully custom skill implementations.
     /// </summary>
     public abstract class Skill
     {
-        /// <summary>
-        /// Strain values are multiplied by this number for the given skill. Used to balance the value of different skills between each other.
-        /// </summary>
-        protected abstract double SkillMultiplier { get; }
-
-        /// <summary>
-        /// Determines how quickly strain decays for the given skill.
-        /// For example a value of 0.15 indicates that strain decays to 15% of its original value in one second.
-        /// </summary>
-        protected abstract double StrainDecayBase { get; }
-
-        /// <summary>
-        /// The weight by which each strain value decays.
-        /// </summary>
-        protected virtual double DecayWeight => 0.9;
-
         /// <summary>
         /// <see cref="DifficultyHitObject"/>s that were processed previously. They can affect the strain values of the following objects.
         /// </summary>
         protected readonly LimitedCapacityStack<DifficultyHitObject> Previous = new LimitedCapacityStack<DifficultyHitObject>(2); // Contained objects not used yet
 
-        /// <summary>
-        /// The current strain level.
-        /// </summary>
-        protected double CurrentStrain { get; private set; } = 1;
-
-        /// <summary>
-        /// The length of each strain section.
-        /// </summary>
-        protected virtual int SectionLength => 400;
-
+        
         /// <summary>
         /// Mods for use in skill calculations.
         /// </summary>
         protected IReadOnlyList<Mod> Mods => mods;
 
-        private double currentSectionPeak = 1; // We also keep track of the peak strain level in the current section.
-
-        private double currentSectionEnd;
-
-        private readonly List<double> strainPeaks = new List<double>();
-
         private readonly Mod[] mods;
-
         protected Skill(Mod[] mods)
         {
             this.mods = mods;
         }
 
         /// <summary>
-        /// Process a <see cref="DifficultyHitObject"/> and update current strain values accordingly.
+        /// Process a <see cref="DifficultyHitObject"/>.
         /// </summary>
-        public void Process(DifficultyHitObject current)
+        /// <param name="current">The <see cref="DifficultyHitObject"/> to process.</param>
+        public virtual void Process(DifficultyHitObject current)
```

Comment:
I'm a little against this method being `virtual` if it's dong `Previous.Push(current)` as its base implementation. As it is, it means every implementation _has_ to `base.Process(current)` at the end of the method otherwise it's just wrong, right?

Would suggest something like...

```csharp
internal void ProcessInternal(DifficultyHitObject current)
{
    Process(current);
    Previous.Add(current);
}

protected abstract void Process(DifficultyHitObject current);
```

Reply:
From what I recall the argument for this was that a theoretical consumer might want to, say, store *all* historical objects rather than use the limited stack. That may be better written in the way you have it, fwiw. I'm still not overly sold on letting consumers keep full history but I don't want to argue against that any more.

Reply:
The next PR up (https://github.com/ppy/osu/pull/11688) basically just exposes a property to set the history length, which is done in the ctor. I think the intention is to still use `Previous` and my proposal doesn't change that - only makes it so that `Previous` actually always means previous.

Reply:
Oh right, I forgot that was changed again. I think this is just stale code then.

Reply:
Ah yea, that's better.
Fixed in 5bdd15f7468a2319a3370850d966458e0936b232